crates/adapters/src/transport/kafka/ft/input.rs

```diff
@@ -99,132 +102,215 @@ impl ClientContext for KafkaFtInputContext {
-                                Offset::Offset(offsets.end),
-                            )
-                            .map_err(|error| self.refine_error(error).1)?;
-                        buffered_partition.push(MessageBuffer::new(offsets.end));
-                    }
-                    buffered_messages.push(buffered_partition);
+        let initial_offsets = match command_receiver.blocking_recv_seek()? {
+            Some(metadata) => metadata
+                .parse(n_partitions)?
+                .into_iter()
+                .map(|range| Offset::Offset(range.end))
+                .collect::<Vec<_>>(),
+            None => match &config.start_from {
+                KafkaStartFromConfig::Earliest => {
+                    iter::repeat_n(Offset::Beginning, n_partitions).collect()
                 }
-            }
-            None => {
-                for (topic, n_partitions) in topics.iter().zip(partition_counts.iter()) {
-                    for partition in 0..*n_partitions {
-                        assignment
-                            .add_partition_offset(
-                                topic.as_str(),
-                                partition as i32,
-                                Offset::Beginning,
-                            )
-                            .map_err(|error| self.refine_error(error).1)?;
+                KafkaStartFromConfig::Latest => iter::repeat_n(Offset::End, n_partitions).collect(),
+                KafkaStartFromConfig::Offsets(offsets) => {
+                    for (offset, partition) in offsets.iter().copied().zip(0..) {
+                        let (low, high) = self
+                            .kafka_consumer
+                            .fetch_watermarks(topic, partition, METADATA_TIMEOUT)
+                            .map_err(|e| {
+                                anyhow!("error fetching metadata for partition '{partition}': {e}",)
+                            })?;
+
+                        // Return an error if the specified offset doesn't exist.
+                        if !(low..=high).contains(&offset) {
+                            bail!("configuration error: provided offset '{offset}' not currently in partition '{partition}'");
+                        }
                     }
-                    buffered_messages.push(
-                        iter::repeat_with(|| MessageBuffer::new(0))
-                            .take(*n_partitions)
-                            .collect(),
-                    );
+
+                    offsets.iter().copied().map(Offset::Offset).collect()
                 }
-            }
+            },
         };
+        let next_offsets = initial_offsets
+            .iter()
+            .map(|o| match o {
+                Offset::Offset(offset) => *offset,
+                _ => 0,
+            })
+            .collect::<Vec<_>>();
+
+        let mut assignment = TopicPartitionList::new();
+        for (partition, offset) in (0..).zip(initial_offsets) {
```

Comment:
Do you need to check if there are more or fewer partitions in the topic than there are initial offsets in the connector config?

Reply:
I should do that, yeah, oops.